tools/tpm2_nvread.c

```diff
@@ -158,8 +158,8 @@ static bool nv_read(TSS2_SYS_CONTEXT *sapi_context) {
         ctx.size_to_read -= nv_data.t.size;
         ctx.offset += nv_data.t.size;
 
-        memcpy(data_buffer, nv_data.t.buffer, data_offest);
-        data_offest += nv_data.t.size;
+        memcpy(data_buffer + data_offset, nv_data.t.buffer, nv_data.t.size);
+        data_offset += nv_data.t.size;
     }
 
     hexdump(data_buffer, data_size);
```

Comment:
I think there's a bug here too, shouldn't this be?

```
hexdump(data_buffer, data_offset);
```

Also, do you know why there's a data_offset and ctx.offset? It seems the state is duplicated or am I missing something?

Reply:
data_offset == data_size at the end of the loop.

Reply:
@williamcroberts well, that's not true on my test:

```
$ echo -n foobar > data.txt

$ hexdump -C -v data.txt 
00000000  66 6f 6f 62 61 72                                 |foobar|
00000006

$ tpm2_nvwrite -x 0x1500018 -a 0x40000001 -f data.txt
66 6f 6f 62 61 72 

$ tpm2_nvread -x 0x1500018 -a 0x40000001 -s 6 -o 0
data_size 32 ctx.size_to_read 0 data_offset 6
000000: 66 6f 6f 62 61 72 00 00 c8 ad 07 91 78 7f 00 00  foobar......x...
000010: 10 60 b2 e9 92 55 00 00 05 00 00 00 00 00 00 00  .`...U..........
```

When hexdump is called, data_size = 32 and data_offset = 6. That's why 32 bytes are printed instead of just 6.

The problem is that nv_public.t.nvPublic.dataSize > nv_data.t.size so that's why I suggested that change, that way you print the actual number of bytes read. After using data_offset instead of data_size:

```
$ tpm2_nvread -x 0x1500018 -a 0x40000001 -s 6 -o 0                    
000000: 66 6f 6f 62 61 72                                foobar
```